### ISSUE TITLE
Null in DbContext.SaveChanges() -> DetectChanges() ...

### ISSUE THREAD

**Issue by RheaAyase** [dotnet/efcore#9852]
Calling SaveChanges() throws a random null reference exception. 

```
Exception message:
   Object reference not set to an instance of an object.
Stack trace:
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.<>c.<get_Entries>b__38_1(InternalEntityEntry e)
   at System.Linq.Utilities.<>c__DisplayClass1_0`1.<CombinePredicates>b__0(TSource x)
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.ChangeDetector.DetectChanges(IStateManager stateManager)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
```

### Steps to reproduce
It's very random, maybe 1 in 100. I don't understand why or when does it happen. There is a bunch of stuff added (new rows) and likely a bunch of stuff modified. Sometimes it may be called without any changes, relying on the DetectChanges to do it's job.

### Further technical details
EF Core version: 2.0.0
Database Provider: Pomelo.EntityFramework.Core.MySql 2.0.0-rtm-10062
Operating system: Fedora 26 Linux 4.11.11-300.fc26.x86_64
.NET Core SDK: https://copr.fedorainfracloud.org/coprs/g/dotnet-sig/dotnet/ (the same source as RHEL packages)
CLI `dotnet run`

**Comment by RheaAyase** [comment]
Update: As soon as it hits this exception, all the following attempts to SaveChanges() on the same context will fail as well. Workaround is to throw away the context and create a new one, throwing away all the unsaved changes with it.

**Comment by ajcvickers** [comment]
@RheaAyase There isn't enough info her for us to be able to reproduce this. Could you post a complete code listing or project that demonstrates what you are seeing?

**Comment by RheaAyase** [comment]
Well you can see [the code](https://github.com/RheaAyase/Botwinder.discord/tree/dev), but you won't be able to test it without knowing a whole lot about what is it doing and the technology it's communicating with. And as I said it's extremely rare and random. I insert 200 new rows and modify a bunch more and it won't happen, then randomly the next time I run it, it's there right from the start.

TLDR for the code:
* Does all kinds of changes to the dbContext which is saved in an "update loop" running on fixed "framerate," such as logging requests and exceptions, or storing user data.
* All the "reads" (queries, loops, any operation with the data) spawn their own context.
* Database is [not created by EFCore](https://github.com/RheaAyase/Botwinder.discord/blob/dev/db_botwinder.sql)

**Comment by ajcvickers** [comment]
@RheaAyase I took a quick look at the code and didn't see anything obvious. Two things to confirm are:
* Context instances are not accessed by multiple threads concurrently. (DbContext and related types are not thread-safe.)
* Any async operations on a context are awaited before a new operation is started. (Sometimes WhenAll can be used, but figuring out when it is safe (i.e. will not cause concurrent access) and when it is not safe is difficult, so it's best not to do this.

Beyond this, without a way to reproduce the issue there is not much more we can do here.

**Comment by RheaAyase** [comment]
> Context instances are not accessed by multiple threads concurrently. (DbContext and related types are not thread-safe.)

Yup, saving should happen in just two places (one when it's connecting, and another in the update) and is also locked to be safe...
Queries always instantiate their own context.

> Beyond this, without a way to reproduce the issue there is not much more we can do here.

It shouldn't throw a null reference exception or anything alike **inside** of that code, regardless of the cause. Something is not quite safe in there.
